YNU-354: Protocol Docs (#69)
layer-3/docs: docs/protocol/off-chain/overview.mdx

---
sidebar_position: 1
title: Off-Chain RPC Overview
---

import Tooltip from '@site/src/components/Tooltip';

# Off-Chain RPC Protocol Overview

The Off-Chain RPC Protocol defines how clients communicate with <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip> to perform state channel operations without touching the blockchain.

---

## What is Nitro RPC?

> **Nitro RPC** is a lightweight, WebSocket-based RPC protocol designed for <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">state channel</Tooltip> communication. It uses a compact JSON array format for efficiency and includes signature-based authentication.

Reviewer: ```suggestion
**Nitro RPC** is a lightweight RPC protocol designed for <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">state channel</Tooltip> communication. It uses a compact JSON array format for efficiency and includes signature-based authentication.
```

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

:::

### NitroRPC/0.2 (Legacy)

**Features**:
- Basic state updates for <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain">app sessions</Tooltip>
- All updates redistribute existing funds
- Cannot add or remove funds from active sessions
- Must close and recreate sessions to change total funds

**Use Case**: Maintained for existing applications, not recommended for new development.

### NitroRPC/0.4 (Current)

**Features**:
- Intent-based state updates: **OPERATE**, **DEPOSIT**, **WITHDRAW**
- Add funds to active <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain">app sessions</Tooltip> (DEPOSIT)
- Remove funds from active sessions (WITHDRAW)
- Better error handling and validation
- Enhanced security checks

**Use Case**: All new implementations should use this version.

---

## Communication Architecture

> Nitro RPC operates over WebSocket connections, enabling bidirectional real-time communication between clients and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip>.

Reviewer: ```suggestion
Nitro RPC enables bidirectional real-time communication between clients and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip>.
```

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

```mermaid
graph TB

>     A[Client Application] <-->|WebSocket Connection| B[Clearnode]

Reviewer: We shouldn't have it here as it depends on the protocol chosen by implementation

Author: Can we make it RPC connection?

Reviewer: Yep

Author: [8c9e9266b03098b33f5e167595d73a24964b0c12](https://github.com/layer-3/docs/pull/69/commits/8c9e9266b03098b33f5e167595d73a24964b0c12)

    B <-->|Event Monitoring| C[Blockchain]
    B <-->|State Management| D[Database]
    
    A -->|"1. RPC Request<br/>(signed)"| B
    B -->|"2. Process & Validate"| D
    B -->|"3. RPC Response<br/>(signed)"| A
    
    B -.->|"Async Notifications"| A
    C -.->|"Blockchain Events"| B
    
    style A fill:#e1f5ff
    style B fill:#ffe1f5
    style C fill:#f5ffe1
    style D fill:#fff5e1
```

### Connection Flow

1. **Client Establishes WebSocket**: Open persistent connection to <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip>
2. **Authentication**: Complete 3-step auth flow (auth_request → auth_challenge → auth_verify)
3. **RPC Communication**: Send requests, receive responses
4. **Notifications**: Receive real-time updates (balance changes, <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> events)

> 5. **Keep-Alive**: Periodic ping/pong to maintain connection

Reviewer: We can mark it is optional and say that it depends on the implementation chosen

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

---

## Message Categories

Nitro RPC methods are organized into functional categories:

### 1. Authentication Methods

Establish and manage authenticated sessions:
- `auth_request` - Initiate authentication

> - `auth_challenge` - Receive challenge from <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users">a clearnode</Tooltip>

Reviewer: I'd say that in the current protocol implementation, when you submit a request with `auth_request` method, then on success you receive a response with `auth_challenge` method

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

- `auth_verify` - Complete authentication

### 2. Channel Management Methods

Create and manage payment channels:
- `create_channel` - Open new <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip>
- `close_channel` - Cooperatively close <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip>

> - `resize_channel` - Adjust <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates">channel</Tooltip> capacity

Reviewer: I'm not sure, but I'd use _allocations_ instead of _capacity_

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

### 3. Transfer Methods

Move funds between users:
- `transfer` - Send funds off-chain with instant settlement

### 4. App Session Methods

Manage multi-party application channels:
- `create_app_session` - Create new <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain">app session</Tooltip>
- `submit_app_state` - Update session state (with intents)
- `close_app_session` - Finalize and distribute funds

### 5. Query Methods

Read state and configuration:
- `get_config`, `get_assets`, `get_channels`, `get_app_sessions`

> - `get_balance`, `get_ledger_entries`, `get_ledger_transactions`

Reviewer: ```suggestion
- `get_ledger_balances`, `get_ledger_entries`, `get_ledger_transactions`
```

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

---

## Security Model

The Off-Chain RPC Protocol provides multiple layers of security:

### Cryptographic Security

- **ECDSA Signatures**: Every message signed with secp256k1
- **Keccak256 Hashing**: Message integrity verification
- **Challenge-Response Auth**: Prove key ownership without exposing private keys

### Protocol-Level Security

- **Request Ordering**: Timestamps prevent replay attacks
- **Session Expiration**: <Tooltip content="A temporary cryptographic key delegated by a user's main wallet that provides a flexible way for the user to manage security of their funds by giving specific permissions and allowances for specific apps">Session keys</Tooltip> have time limits
- **Spending Allowances**: Limit session key spending power
- **Signature Verification**: All operations require valid signatures

### Network Security

- **WebSocket over TLS**: Encrypted communication channel
- **Origin Validation**: Prevent unauthorized connections

> - **Rate Limiting**: Protection against DoS attacks

Reviewer: I'm not sure whether we need to include the following section as it depends on the implementation

Author: done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)
